gen/zap.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package gen
+
+import (
+	"fmt"
+
+	"go.uber.org/thriftrw/compile"
+)
+
+type zapGenerator struct {
+	mapG  mapGenerator
+	setG  setGenerator
+	listG listGenerator
+}
+
+// zapEncoder returns the Zap type name of the root spec, determining what type
+// the Zap marshaler needs to log it as (i.e. AddString, AppendObject, etc.)
+func (z *zapGenerator) zapEncoder(g Generator, spec compile.TypeSpec) string {
+	root := compile.RootTypeSpec(spec)
+
+	switch t := root.(type) {
+	// Primitives
+	case *compile.BoolSpec:
+		return "Bool"
+	case *compile.I8Spec:
+		return "Int8"
+	case *compile.I16Spec:
+		return "Int16"
+	case *compile.I32Spec:
+		return "Int32"
+	case *compile.I64Spec:
+		return "Int64"
+	case *compile.DoubleSpec:
+		return "Float64"
+	case *compile.StringSpec:
+		return "String"
+	case *compile.BinarySpec:
+		return "String" // encode binary as a string and log as string
+
+	// Containers
+	case *compile.MapSpec:
+		switch t.KeySpec.(type) {
```

Comment:
`compile.RootTypeSpec(t.KeySpec)` because the map key may be a typedef of a
string. If we don't have tests for this, we should add a test case to
typedefs.thrift.